dawn/test/integration-test/unstructured/ToylibIntegrationTestCompareOutput.cpp

```diff
@@ -0,0 +1,343 @@
+#include <generated_verticalSum.hpp>
+TEST(ToylibIntegrationTestCompareOutput, VerticalCopy) {
+  toylib::Grid mesh(32, 32);
+  size_t nb_levels = 5;
+  toylib::FaceData<double> out(mesh, nb_levels);
+  toylib::FaceData<double> in(mesh, nb_levels);
+
+  double initValue = 10.;
+  InitData(mesh.faces(), in, nb_levels, initValue);
+  InitData(mesh.faces(), out, nb_levels, -1.0);
+
+  dawn_generated::cxxnaiveico::verticalSum<toylibInterface::toylibTag>(
+      mesh, static_cast<int>(nb_levels), in, out)
+      .run();
+
+  for(size_t level = 1; level < nb_levels - 1; level++) {
+    for(const auto& f : mesh.faces()) {
+      ASSERT_EQ(out(f, level), 2 * initValue);
+    }
+  }
+}
+
+#include <generated_accumulateEdgeToCell.hpp>
+TEST(ToylibIntegrationTestCompareOutput, Accumulate) {
+  toylib::Grid mesh(32, 32);
+  size_t nb_levels = 1;
+  toylib::EdgeData<double> in(mesh, nb_levels);
+  toylib::FaceData<double> out(mesh, nb_levels);
+
+  InitData(mesh.edges(), in, nb_levels, 1.0);
+  InitData(mesh.faces(), out, nb_levels, -1.0);
+
+  dawn_generated::cxxnaiveico::accumulateEdgeToCell<toylibInterface::toylibTag>(
+      mesh, static_cast<int>(nb_levels), in, out)
+      .run();
+
+  for(const auto& f : mesh.faces())
+    ASSERT_EQ(out(f, 0), 3.0);
+}
+
+#include <generated_diffusion.hpp>
+#include <reference_diffusion.hpp>
+TEST(ToylibIntegrationTestCompareOutput, Diffusion) {
+  toylib::Grid mesh(32, 32, false, 1., 1.);
+  size_t nb_levels = 1;
+
+  toylib::FaceData<double> in_ref(mesh, nb_levels);
+  toylib::FaceData<double> out_ref(mesh, nb_levels);
+  toylib::FaceData<double> in_gen(mesh, nb_levels);
+  toylib::FaceData<double> out_gen(mesh, nb_levels);
+
+  for(const auto& cell : mesh.faces()) {
+    auto [x, y] = cellMidpoint(cell);
+    bool inX = x > 0.375 && x < 0.625;
+    bool inY = y > 0.375 && y < 0.625;
+    in_ref(cell, 0) = (inX && inY) ? 1 : 0;
+    in_gen(cell, 0) = (inX && inY) ? 1 : 0;
+  }
+
+  for(int i = 0; i < 5; ++i) {
+    // Run the stencils
+    dawn_generated::cxxnaiveico::reference_diffusion<toylibInterface::toylibTag>(
+        mesh, static_cast<int>(nb_levels), in_ref, out_ref)
+        .run();
+    dawn_generated::cxxnaiveico::diffusion<toylibInterface::toylibTag>(
+        mesh, static_cast<int>(nb_levels), in_gen, out_gen)
+        .run();
+
+    // Swap in and out
+    using std::swap;
+    swap(in_ref, out_ref);
+    swap(in_gen, out_gen);
+  }
+
+  {
+    AtlasVerifier v;
+    EXPECT_TRUE(v.compareToylibField(mesh.faces(), in_ref, in_gen, nb_levels))
+        << "while comparing output (on cells)";
+  }
+}
+
+#include <generated_gradient.hpp>
+#include <reference_gradient.hpp>
+TEST(ToylibIntegrationTestCompareOutput, Gradient) {
+  const int numCell = 10;
+  auto mesh =
+      toylib::Grid(numCell, numCell, false, M_PI,
+                   M_PI); // periodic seems bugged in toylib, but boundaries are simply cut off.
+                          // since our boundaries are 0 this seems not too bad of a compromise
```

Comment:
I didn't get this last part

Reply:
so boundary edges in toylib have only one entry in their cell neighbors list; the cell inside the domain. Hence the domain can be said to be "cut off" there. For the gradient test case we have zero boundaries. So this kinda works out. 

In atlas we have two entries. Instead of deleting one entry I decided to wrap the domain around. But again, we have zero boundaries, so a boundary edge "sees" another zero on the other end of the domain. 

I'm not sure how to explain this better at this point, open for suggestions ;)